Add -snapshot option for snapshotting table state

bruth opened the pull request:
The purpose of this is to support initializing the state of a stream by
snapshotting the initial state of the table and as well as creating
snapshots of the table on regular intervals. This reduces the overhead
for consumers to produce the current state by reading the stream. A
snapshot enables a consumer to start at the last known snapshot and read
from there rather than the beginning of the stream.

Signed-off-by: Byron Ruth <b@devel.io>

bruth committed:
Add -snapshot option for snapshotting table state

The purpose of this is to support initializing the state of a stream by
snapshotting the initial state of the table and as well as creating
snapshots of the table on regular intervals. This reduces the overhead
for consumers to produce the current state by reading the stream. A
snapshot enables a consumer to start at the last known snapshot and read
from there rather than the beginning of the stream.

Signed-off-by: Byron Ruth <b@devel.io>

bruth commented:
This also adds a `time` value to events (sort of important for an event!). Here is a sample usage:

```bash
% diff-table -csv1 example/file1.csv -key id -snapshot | jq .
{
  "type": "row-stored",
  "time": 1520119177,
  "offset": 1,
  "key": {
    "id": "1"
  },
  "data": {
    "color": "Blue",
    "gender": "Male",
    "id": "1",
    "name": "John"
  }
}
{
  "type": "row-stored",
  "time": 1520119177,
  "offset": 2,
  "key": {
    "id": "2"
  },
  "data": {
    "color": "Red",
    "gender": "Female",
    "id": "2",
    "name": "Pam"
  }
}
{
  "type": "row-stored",
  "time": 1520119177,
  "offset": 3,
  "key": {
    "id": "3"
  },
  "data": {
    "color": "Yellow",
    "gender": "Female",
    "id": "3",
    "name": "Sam"
  }
}
```

bruth commented:
@aaron0browne when you have a few minutes can you review this.

gracecodes144 commented on README.md:
```diff
@@ -115,7 +117,43 @@ diff-table \
+  "key": {...},
+
+  // Full snapshot of the row. This will be present row row-added events, but
+  // can be optionally included for row-changed and row-removed events.
```
> can be optionally included for row-changed and row-removed events

Is this option already implemented or did I miss it in these changes?

bruth replied:
Good catch! Adding..

bruth replied:
Added a new option to support this.

bruth committed:
Add -data option

If set, this will include the full data of the new row for row-changed
events and the full data of the old row for row-deleted events.

Signed-off-by: Byron Ruth <b@devel.io>
